plugin-server/src/worker/ingestion/person-state.ts

```diff
@@ -137,45 +136,50 @@ export class PersonState {
-        if (isNewPerson) {
-            const properties = excludeProperties ? {} : this.eventProperties['$set'] || {}
-            const propertiesOnce = excludeProperties ? {} : this.eventProperties['$set_once'] || {}
-            // Catch race condition where in between getting and creating, another request already created this user
-            try {
-                const person = await this.createPerson(
-                    this.timestamp,
-                    properties || {},
-                    propertiesOnce || {},
-                    this.teamId,
-                    null,
-                    // :NOTE: This should never be set in this branch, but adding this for logical consistency
-                    this.updateIsIdentified,
-                    this.newUuid,
-                    this.event.uuid,
-                    [this.distinctId]
-                )
-                // :TRICKY: Avoid subsequent queries re-fetching person
-                this.personContainer = this.personContainer.with(person)
-                return true
-            } catch (error) {
-                if (!error.message || !error.message.includes('duplicate key value violates unique constraint')) {
-                    Sentry.captureException(error, {
-                        extra: {
-                            teamId: this.teamId,
-                            distinctId: this.distinctId,
-                            timestamp: this.timestamp,
-                            personUuid: this.newUuid,
-                        },
-                    })
-                }
-            }
+        const properties = excludeProperties ? {} : this.eventProperties['$set'] || {}
+        const propertiesOnce = excludeProperties ? {} : this.eventProperties['$set_once'] || {}
+
+        try {
+            return promiseRetry(() => this.createOrGetPerson(properties, propertiesOnce))
+        } catch (error) {
+            // TODO: figure out if we need to add Sentry here or it's already upstream
+            Sentry.captureException(error, {
+                extra: {
+                    teamId: this.teamId,
+                    distinctId: this.distinctId,
+                    timestamp: this.timestamp,
+                    personUuid: this.newUuid,
+                },
+            })
+            throw error
         }
+    }
 
-        // Person was likely created in-between start-of-processing and now, so ensure that subsequent queries
-        // to fetch person still return the right `person`
-        this.personContainer = this.personContainer.reset()
-        return false
+    // If we created the person we already updated the properties, in that case we return true
+    // If the person already existed we do nothing here, which means properties aren't updated and we return false
+    private async createOrGetPerson(properties: Properties, propertiesOnce: Properties): Promise<boolean> {
```

Comment:
I see how we are dealing with the race condition: by checking the cache we can make sure there are no persons waiting to be created.

However, if we have new properties to set to a person that has not been created yet as it's in the cache, won't that cause us to lose the property updates?

Reply:
> I see how we are dealing with the race condition: by checking the cache we can make sure there are no persons waiting to be created.

we don't really check that after this change (we used to do that before). We now only check if the person already exists. If there's a person in the process of being created we'll go ahead and try to create them, which is fine as that just results in us failing the creation, which then results us checking the person existence again (3x total), which would now maybe say the person already exists. 

> However, if we have new properties to set to a person that has not been created yet as it's in the cache, won't that cause us to lose the property updates?

If we return false, then the code that called this function knows that person wasn't created and person properties still need to be updated.

Reply:
> Which is fine as that just results in us failing the creation, which then results us checking the person existence again (3x total), which would now maybe say the person already exists

Oh wow, I didn't notice this was happening 3x.

As creating a person ultimately fires off an `INSERT INTO` query, wouldn't it make sense to add a conflict action to the `INSERT INTO` and remove everything else? e.g. if the insert fails, our conflict action can be to only update properties. If it works it could cut down the entire thing to one query.

Reply:
we should look into doing this afterwards (you can check out the createPerson function that's a transaction that creates the person and adds the id to person distinctID table) - it's a bit tricky because properties is a JSON column and that can get complex to add fields there.

Reply:
Retrying a bit sounds better than straight up failing and I'm all for trying this out later, it could be well out of scope for now, thanks for the replies!